### DISCUSSION TITLE
Use ... to indicate default columns for Progress

### DISCUSSION THREAD

**Discussion by ptmcg** [Textualize/rich#1867] [Ideas]
rich.progress.Progress uses these columns as default if no columns are specified:

        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        TimeRemainingColumn(),

If I just want to specify some extensions, I have to give the whole list of columns, including the defaults. For instance, to add a Spinner before and a TimeElapsed after, I have to do this:

    # current Progress construction
    # add spinner before and elapsed after default columns
    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        TimeRemainingColumn(),
        TimeElapsedColumn(),
        console=console,
        transient=True,
    ) as progress:

Wouldn't it be a friendlier API to accept `...` as a value indicating "default columns here", so that this same code could be written as:

    # Progress construction using ... as placeholder for the default columns
    # add spinner before and elapsed after default columns
    with Progress(
        SpinnerColumn(),
        ...,
        TimeElapsedColumn(),
        console=console,
        transient=True,
    ) as progress:

Or as a more conventional option, define the default columns as a classmethod:

    # Progress construction using ... as placeholder for the default columns
    # add spinner before and elapsed after default columns
    with Progress(
        SpinnerColumn(),
        *Progress.default_columns(),
        TimeElapsedColumn(),
        console=console,
        transient=True,
    ) as progress:

Just so I don't have to hard-code the default columns in my own code.

**Comment by willmcgugan** [comment]
Hi Paul,

Yeah, I think this is a good idea.

**Reply by willmcgugan** [reply]
https://github.com/Textualize/rich/issues/1891

**Comment by ptmcg** [comment]
After writing that post, I think I'll just do a PR that adds the classmethod. "..." as a parameter might be sexy, but it defies typing, so I take that as a sign.

**Reply by willmcgugan** [reply]
Fair enough. Would it not accept the arguments as a Union of Column and EllipsisType ?

**Reply by ptmcg** [reply]
I tried, but `Ellipsis` is not a valid type for typing, it turns out, nor is `type(Ellipsis)`.

**Reply by willmcgugan** [reply]
Interesting, looks like there is an EllipsisType in 3.10 https://docs.python.org/3/library/types.html#types.EllipsisType

Shame, it was sexy. Guess classmethod will have to do.

**Reply by ptmcg** [reply]
Darn that "compatibility with older Python versions" thing!

**Comment by ptmcg** [comment]
Hrmmmm, PR failed to build. "Works on my machine."

**Reply by willmcgugan** [reply]
Weird error on the Windows CI. Doesn't look related to your change. Will re-run jobs in case it was a glitch.

**Comment by ptmcg** [comment]
PR merged - next release will have get_default_columns() classmethod on Progress.
